src/rust/engine/src/watch.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::collections::HashSet;
+use std::path::PathBuf;
+use std::sync::{Arc, Weak};
+use std::thread;
+use std::time::Duration;
+
+use boxfuture::{BoxFuture, Boxable};
+use crossbeam_channel::{self, Receiver, RecvTimeoutError, TryRecvError};
+use futures01 as futures;
+use futures01::future::Future;
+use log::{error, warn};
+use notify::{RecommendedWatcher, RecursiveMode, Result as NotifyResult, Watcher};
+use parking_lot::Mutex;
+
+use graph::{Graph, InvalidationResult};
+use logging;
+
+use crate::nodes::NodeKey;
+
+///
+/// An InvalidationWatcher maintains a Thread that receives events from a notify Watcher.
+///
+/// If the spawned Thread exits for any reason, InvalidationWatcher::running() will return False,
+/// and the caller should create a new InvalidationWatcher (or shut down, in some cases). Generally
+/// this will mean polling.
+///
+/// TODO: Need the above polling, and need to make the watch method async.
```

Comment:
The `watch` method is now async, but should ensure that we have something monitoring whether the liveness thread has died.

Reply:
Is the liveness check something that  belongs in this PR?

Reply:
based on your comment [here](https://github.com/pantsbuild/pants/issues/4999#issuecomment-564815378) I would like to add the liveness polling in a follow up PR to keep this concise.